service/worker/processor.go

```diff
@@ -0,0 +1,179 @@
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package worker
+
+import (
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"encoding/json"
+	"github.com/uber-common/bark"
+	"github.com/uber-go/kafka-client/kafka"
+	"github.com/uber/cadence/.gen/go/replicator"
+	"github.com/uber/cadence/common"
+	"github.com/uber/cadence/common/logging"
+	"github.com/uber/cadence/common/messaging"
+)
+
+type (
+	replicationTaskProcessor struct {
+		topicName    string
+		consumerName string
+		client       messaging.Client
+		consumer     kafka.Consumer
+		isStarted    int32
+		isStopped    int32
+		shutdownWG   sync.WaitGroup
+		shutdownCh   chan struct{}
+		config       *Config
+		logger       bark.Logger
+	}
+)
+
+func newReplicationTaskProcessor(topic, consumer string, client messaging.Client, config *Config,
+	logger bark.Logger) *replicationTaskProcessor {
+	return &replicationTaskProcessor{
+		topicName:    topic,
+		consumerName: consumer,
+		client:       client,
+		shutdownCh:   make(chan struct{}),
+		config:       config,
+		logger: logger.WithFields(bark.Fields{
+			logging.TagWorkflowComponent: logging.TagValueReplicationTaskProcessorComponent,
+			logging.TagTopicName:         topic,
+			logging.TagConsumerName:      consumer,
+		}),
+	}
+}
+
+func (p *replicationTaskProcessor) Start() error {
+	if !atomic.CompareAndSwapInt32(&p.isStarted, 0, 1) {
+		return nil
+	}
+
+	logging.LogReplicationTaskProcessorStartingEvent(p.logger)
+	consumer, err := p.client.NewConsumer(p.topicName, p.consumerName, p.config.ReplicatorConcurrency)
+	if err != nil {
+		logging.LogReplicationTaskProcessorStartFailedEvent(p.logger, err)
+		return err
+	}
+
+	if err := consumer.Start(); err != nil {
+		logging.LogReplicationTaskProcessorStartFailedEvent(p.logger, err)
+		return err
+	}
+
+	p.consumer = consumer
+	p.shutdownWG.Add(1)
+	go p.processorPump()
+
+	logging.LogReplicationTaskProcessorStartedEvent(p.logger)
+	return nil
+}
+
+func (p *replicationTaskProcessor) Stop() {
+	if !atomic.CompareAndSwapInt32(&p.isStopped, 0, 1) {
+		return
+	}
+
+	logging.LogReplicationTaskProcessorShuttingDownEvent(p.logger)
+	defer logging.LogReplicationTaskProcessorShutdownEvent(p.logger)
+
+	if atomic.LoadInt32(&p.isStarted) == 1 {
+		close(p.shutdownCh)
+	}
+
+	if success := common.AwaitWaitGroup(&p.shutdownWG, time.Minute); !success {
+		logging.LogReplicationTaskProcessorShutdownTimedoutEvent(p.logger)
+	}
+}
+
+func (p *replicationTaskProcessor) processorPump() {
+	defer p.shutdownWG.Done()
+
+	var workerWG sync.WaitGroup
+	for i := 0; i < p.config.ReplicatorConcurrency; i++ {
+		workerWG.Add(1)
+		go p.worker(&workerWG)
+	}
+
+processorPumpLoop:
```

Comment:
is this loop necessary?

Reply:
You are right, it is not necessary.  Previously I had more case statements which I removed later but forgot to remove the loop.